[cyiting] iP

cyiting opened the pull request:
## Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use
All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
And it is **FREE**!

Features:

- [ ] Managing tasks
- [ ] Managing deadlines 
- [ ] Managing events
- [ ] Reminders (coming soon)
If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

cyiting committed:
created simple duke

cyiting committed:
added ability to save tasks in list

cyiting committed:
implemented TaskList and Task as classes
added functionality of marking and unmarking tasks

cyiting committed:
added Deadline, Event, ToDo as subclasses of Task
converted Task to abstract class

cyiting committed:
included automated text ui testing
updated tab to 3 spaces when printing added/marked tasks

cyiting committed:
Modified Duke to handle exceptions
Added DukeException class and two subclasses EmptyTaskDescriptionException and InvalidCommandException

cyiting committed:
Added support for deleting task from list

cyiting committed:
Add ability to save task in local device in the directory data

cyiting committed:
Change Dates and Times Features

Task classes recognise datetime as Strings

cyiting committed:
Merge branch 'branch-level-7'

cyiting committed:
Merge branch 'branch-level-8' and resolve conflicts

DateTime format is incompatible in branch-level-8

cyiting committed:
create classes for Parser, Storage and Ui

Dealing with interactions with user, parsing information and sorting files occur in Duke.

Let's extract methods to separate classes and OOP Duke.

cyiting committed:
Organise classes into packages

Create 3 packages, duke, duke.exception, duke.task

cyiting committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

cyiting committed:
Set up gradle to run Duke using gradle

cyiting committed:
Add JUnit tests to test behaviour of code

Include 3 test classes for ToDo, Deadline, Event

cyiting committed:
Add JavaDoc comments to all public methods.

cyiting committed:
Tweak code to comply with coding standard

cyiting committed:
Change code to comply with coding standard

cyiting committed:
Add functionality to find tasks using a keyword

User types "find" followed by a keyword and a list of tasks containing the
keyword is shown.

cyiting committed:
Merge branch 'branch-a-codingstandard'

Resolve conflicts in TaskList.java and Task.java.

cyiting committed:
Merge branch 'branch-level-9'

SHni99 commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,21 @@
+package duke.exception;
+public class DukeException extends Exception {
+    String errorMessage;
```
Can private this variable

SHni99 reviewed (commented):
LGTM. Good coding standard. Code flow is easy to understand and pleasing to read. Author followed coding standards and coding design which is almost mistake free. keep it up!

cyiting committed:
Add a GUI to Duke

Create a GUI using JavaFX that takes in user input and responds. Duke lacks
functionality to print greetings message when application is first launched.

SHni99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+        tasks = new ArrayList<>();
+    }
+
+    void addTask(Task task) {
```
Add javadoc for each methods in every classes will be more consistent :)

SHni99 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    /**
+     * @inheritDoc
```
good use of inheritDoc to document inherited methods

ChickenChiang commented on src/main/java/duke/task/Event.java:
```diff
@@ -6,24 +6,43 @@ public class Event extends Task {
     LocalDateTime end;
 
+    /**
+     * Constructor for Event.
```
Perhaps leave a line between method description and param fields in the documentation for better readability?

cyiting committed:
Make minor changes to Duke

cyiting committed:
Include JavaDoc comments for all classes and InvalidTaskNumberException

cyiting committed:
Make changes to GUI layout

cyiting committed:
Update GUI to display greetings message whenever launched

cyiting committed:
Add Command classes for all types of commands from user.

Include JavaDoc commands for classes and public methods.

cyiting committed:
Convert for loops into Streams

There are several methods that contain for loops.

cyiting committed:
Add assertions to test code

No checks are done to ensure that a response is formed when user enters input or that
a file path exists

Asserting that file path exists ensures that creating a File object with the file path will not
throw an error. Asserting that the response is not an empty String ensures that there is no
error in parsing user commands and having Duke respond to the command.

Using assertions is desirable because it can indicate any errors made by the coder.

cyiting committed:
TaskList class: extract a method checkTaskNumValidity

TaskList methods (e.g. markTask, unmarkTask, deleteTask) have some common
behaviours.

The common behaviors across taskList methods cause code duplication.

Extracting the common behavior into a method allows us to use the same
method for checking the validity of a task number when new commands are added.

Let's pull up behaviors common to all methods into a new private method called
checkTaskNumValidity.

cyiting committed:
Change Duke to deal with exceptions concerning date-time

AddDeadlineCommand, AddEventCommand, AddTodoCommand is unable to deal with exceptions
regarding date-time.

Create an InvalidDateTimeException and InvalidTaskDescription Exception. All methods that involve
creating Deadlines and Events now check if the date-time input by the user is in the correct format,
verify that the start date is before the end date and that there are no missing date or time inputs.

Now Duke will feedback to the user if there is a date-time related exception.

cyiting committed:
Merge pull request #1 from cyiting/branch-a-assertions

Branch-a-assertions

cyiting committed:
Merge branch 'branch-a-codequality'

cyiting committed:
Merge branch 'branch-a-streams'

Resolve merge conflicts in src/main/java/duke/task/Task.java

cyiting committed:
Create Tag, EmptyTag, TagList classes to allow users to tag their tasks

Implementation of tagging is not done.

cyiting committed:
Add functionality for the user to add a tag to a task when creating it

The tasks in the task list cannot be categorised. This functionality allows users to categorise their
tasks by the tag name.

Adding an attribute of type Tag to the abstract class Task allows the tag to be displayed when the
tasks are listed and the users are now able to get a list of all the tasks under the same tag.

cyiting committed:
Change the icon to be a circle instead of a square

cyiting committed:
Add user guide for Duke

cyiting committed:
Fill background of HBox

cyiting committed:
Include variable arguments for ListCommand's constructor

Instead of having 2 separate Command classes to execute listing all tasks
and listing tasks tagged to a keyword, variable arguments allow them to be
combined into a class.

cyiting committed:
Change Duke to deal with more exceptions

cyiting committed:
Remove comments in DialogBox class

cyiting committed:
Merge branch 'branch-a-bettergui'

cyiting committed:
Convert all user input to lower case before processing it

cyiting committed:
Update user guide.

cyiting committed:
Add product screenshot

cyiting committed:
Update product screenshot

cyiting committed:
Update product screenshot

cyiting committed:
Edit GUI to expand with the size of the window
